packages/flame/lib/src/components/spawn_component.dart

```diff
@@ -0,0 +1,107 @@
+import 'dart:async';
+import 'dart:math';
+
+import 'package:flame/components.dart';
+import 'package:flame/effects.dart';
+import 'package:flame/experimental.dart';
+
+/// The [SpawnComponent] is a non-visual component which can spawn
+/// [PositionComponent]s randomly within a set [area]. If [area] is not set it
+/// will use the size of the nearest ancestor that provides a size.
+///
+/// If you want your components to spawn with a static period, just set [period]
+/// and if you want them to spawn within a random period set [minPeriod] and
+/// [maxPeriod].
+class SpawnComponent extends Component {
+  SpawnComponent({
+    required this.factory,
+    this.area,
+    this.within = true,
+    double? period,
+    double? minPeriod,
+    double? maxPeriod,
+    Random? random,
+    super.key,
+  })  : _period = period ?? 0,
+        _random = random ?? _randomFallback,
+        assert(
+          period != null || (minPeriod != null && maxPeriod != null),
+          'Either the static period must be defined, or both minPeriod and '
+          'maxPeriod must be defined',
+        );
+
+  /// The function used to create new components to spawn.
+  PositionComponent Function() factory;
```

Comment:
maybe would be cool to receive the index of the created component, maybe people could do some logic on top of it, like, spawn X is even and Y if odd, or every 10 spawnmed, spawn a boss.

Reply:
It loses the nice way of being able to do `factory: MyComponent.new`, but it is true that the index could be useful, so I'll add it!

Reply:
yeah, it is a shame that we don't have unions yet :(

Reply:
For this use-case unions would be really dirty since it then would have to check what the function signature is in runtime to be able to either send or not send the argument